File: ping_oc.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <arpa/inet.h>
#include <sys/socket.h>

#define MAX_BUF_SIZE 1024

int main() {
    int client_socket;
    struct sockaddr_in server_addr;
    char message[MAX_BUF_SIZE];
    int port;
    char ip[16];

    // Solicitar al usuario ingresar la dirección IP del servidor
    printf("Ingrese la dirección IP del servidor: ");
    scanf("%s", ip);

    // Solicitar al usuario ingresar el puerto del servidor
    printf("Ingrese el puerto del servidor: ");
    scanf("%d", &port);

    // Crear el socket TCP
    client_socket = socket(AF_INET, SOCK_STREAM, 0);
    if (client_socket < 0) {
        perror("Error al crear el socket del cliente");
        exit(EXIT_FAILURE);
    }

    // Configurar la dirección del servidor
    memset(&server_addr, 0, sizeof(server_addr));
    server_addr.sin_family = AF_INET;
    server_addr.sin_port = htons(port);
    server_addr.sin_addr.s_addr = inet_addr(ip);

    // Conectar al servidor
    if (connect(client_socket, (struct sockaddr*)&server_addr, sizeof(server_addr)) < 0) {
        perror("Error al conectar con el servidor");
        exit(EXIT_FAILURE);
    }

    printf("Conectado al servidor\n");

    // Enviar y recibir paquetes ICMP al servidor
    int icmp_seq = 0;
    while (1) {
        // Incrementar el número de secuencia ICMP
        icmp_seq++;

        // Construir el mensaje ICMP simulado
        snprintf(message, MAX_BUF_SIZE, "Paquete ICMP_SEQ=%d", icmp_seq);

        // Enviar el mensaje al servidor
        if (send(client_socket, message, strlen(message), 0) < 0) {
            perror("Error al enviar el mensaje al servidor");
            exit(EXIT_FAILURE);
        }

        // Esperar la respuesta del servidor
        memset(message, 0, sizeof(message));
        if (recv(client_socket, message, MAX_BUF_SIZE, 0) <= 0)
        {
        perror("Error al recibir datos del servidor");
        exit(EXIT_FAILURE);
        }   


        // Imprimir la respuesta del servidor
        printf("Respuesta del servidor: %s\n", message);

        // Esperar un tiempo antes de enviar el próximo paquete (opcional)
        sleep(1);
    }

    // Cerrar el socket del cliente
    close(client_socket);

    return 0;
}

File: ping_oc.py
import socket
import time
import argparse

def main():


    # Solicitar al usuario ingresar la dirección IP del servidor y el puerto
    server_ip = input("Ingrese la dirección IP del servidor: ")
    server_port = int(input("Ingrese el puerto del servidor: "))

    # Crear un socket TCP/IP
    client_socket = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

    # Conectar el socket al servidor
    client_socket.connect((server_ip, server_port))

    print(f"Conectado al servidor {server_ip}:{server_port}")

    # Crear un objeto ArgumentParser
    parser = argparse.ArgumentParser(description="Enviar paquetes ICMP a un servidor.")

    # Añadir el argumento para el número de paquetes
    parser.add_argument("--num-paquetes", type=int, default=float('inf'), help="Número de paquetes a enviar.")

    # Parsear los argumentos
    args = parser.parse_args()

    # Inicializar el número de secuencia ICMP
    seq = 0

while seq < args.num_paquetes:
    # Incrementar el número de secuencia ICMP
    seq += 1

    # Obtener la marca de tiempo antes de enviar el mensaje
    start_time = time.time()

    # Construir el mensaje 
    message = f"{seq}"

    # Enviar el mensaje al servidor
    client_socket.sendall(message.encode("utf-8"))

    # Esperar la respuesta del servidor
    response = client_socket.recv(1024).decode("utf-8")

    # Calcular la diferencia de tiempo después de recibir la respuesta
    end_time = time.time()
    elapsed_time = end_time - start_time

    # Imprimir la respuesta del servidor y el tiempo transcurrido
    print(f"Respuesta del servidor: {response}, Tiempo transcurrido: {elapsed_time} segundos")

    # Esperar un tiempo antes de enviar el próximo paquete (opcional)
    time.sleep(1)

# Cerrar el socket del cliente
client_socket.close()
